### PR TITLE
Add linux and sudo support

### PR DESCRIPTION
**Pull request by silverwind**
Two enhancements here:
- when invoked with `sudo` will try to downgrade to the user who invoked it
- adds default uid for (most) linux platforms

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,8 +1,17 @@
...
 
 module.exports = function () {
 	if (isRoot() && process.setuid) {
-		process.setuid(501);
+		var uid = parseInt(process.env.SUDO_UID, 10);
+		if (uid && uid > 0) {
+			process.setuid(uid);
+		} else {
+			process.setuid(defaultUID[process.platform]);
+		}
```
Comment (sindresorhus) [inline index.js:15]:
``` js
process.setuid(uid && uid > 0 ? uid  : defaultUID[process.platform]);
```

File: index.js [hunk 2]
```diff
@@ -1,8 +1,17 @@
 'use strict';
 var isRoot = require('is-root');
+var defaultUID = {
+	'darwin': 501,
+	'linux': 1000
+};
```
Comment (sindresorhus) [inline index.js:8]: don't quote keys

File: test.js [hunk 3]
```diff
@@ -4,6 +4,10 @@ var downgradeRoot = require('./');
 
 test(function (t) {
 	var _ = process.getuid;
+	var defaultUID = {
+		'darwin': 501,
+		'linux': 1000
+	};
```
Comment (sindresorhus) [inline test.js:12]: don't quote keys

File: index.js [hunk 4]
```diff
@@ -1,8 +1,17 @@
...
 
 module.exports = function () {
 	if (isRoot() && process.setuid) {
-		process.setuid(501);
+		var uid = parseInt(process.env.SUDO_UID, 10);
+		if (uid && uid > 0) {
+			process.setuid(uid);
+		} else {
+			process.setuid(defaultUID[process.platform]);
```
Comment (sindresorhus) [inline index.js]: What default should we use for other platforms?

Windows doesn't count as it doesn't have `process.setuid`, but others like OpenBSD or Solaris.

**Comment by sindresorhus** [comment]
Nice! Thanks for fixing my ignorance around permissions :dancer:

File: index.js (hunk 4 above)
Comment (silverwind) [inline index.js]: I'll research.

**Comment by sindresorhus** [comment]
// @Munter

File: index.js [hunk 5]
```diff
@@ -1,8 +1,17 @@
 'use strict';
 var isRoot = require('is-root');
+var defaultUID = {
```
Comment (sindresorhus) [inline index.js]:
``` js
var DEFAULT_UID = {
```

File: test.js [hunk 6]
```diff
@@ -4,6 +4,10 @@ var downgradeRoot = require('./');
 
 test(function (t) {
 	var _ = process.getuid;
+	var defaultUID = {
```
Comment (sindresorhus) [inline test.js]:
``` js
var DEFAULT_UID = {
```

**Comment by silverwind** [comment]
@sindresorhus All done. Only the windows platform missing, not sure yet what to do there.

**Comment by silverwind** [comment]
@sindresorhus Oh, wait. Let me fix a whitespace issue.

**Comment by silverwind** [comment]
@sindresorhus Ok, merge at will.

File: index.js [hunk 7]
```diff
@@ -1,8 +1,15 @@
 'use strict';
 var isRoot = require('is-root');
+var DEFAULT_UID = {
+	darwin: 501,
+	freebsd: 1000,
+	linux: 1000,
+	sunos: 100
+};
```
Comment (sindresorhus) [inline index.js:8]: What should it do when there's a platform not in this list?

We should probably:

``` js
if (isRoot() && process.setuid && process.platform in DEFAULT_UID) {
```

**Comment by sindresorhus** [comment]
> Only the windows platform missing, not sure yet what to do there.

Nothing. Windows doesn't support setting `uid`.

**Comment by silverwind** [comment]
According to the [docs](http://nodejs.org/api/process.html#process_process_platform), it can only be one of the mentioned five values, but let me refactor once more to do nothing on unknown platforms.

**Comment by sindresorhus** [comment]
> According to the docs, it can only be one of the mentioned five values, but let me refactor once more to do nothing on unknown platforms.

Yes, but Node might add more platforms in the future.

**Comment by silverwind** [comment]
Alright, that should do it.

**Comment by sindresorhus** [comment]
Thank you :)

Where did you find the default UIDs? Any resources for where I can read more about this?

**Comment by silverwind** [comment]
Just by googling, really.

For Linux: http://www.linfo.org/uid.html
For Solaris: https://docs.oracle.com/cd/E19120-01/open.solaris/819-2379/userconcept-3/index.html

Particulary interesting part of the first link:

> Some Linux distributions (i.e., versions) begin UIDs for non-privileged users at 100. Others, such as Red Hat, begin them at 500, and still others, such Debian, start them at 1000. Because of the differences among distributions, manual intervention can be necessary if multiple distributions are used in a network  in an organization. 

As far as I know, all common linux distribution use 1000+ nowadays, but there sure are exceptions. I'll check on a RedHat box later.

**Comment by sindresorhus** [comment]
Thanks :)

**Pull request merged**